src/v/archival/archival_policy.h

```diff
@@ -10,60 +10,34 @@
 
 #pragma once
 
-#include "archival/manifest.h"
+#include "archival/types.h"
 #include "model/fundamental.h"
 #include "storage/log_manager.h"
 
 namespace archival {
 
-/// Policy that controls how archiver picks upload candidates
-enum class upload_policy_selector {
-    /// Archive only original, non-compacted segments
-    archive_non_compacted,
+struct upload_candidate {
+    ss::lw_shared_ptr<storage::segment> source;
+    segment_name exposed_name;
+    model::offset starting_offset;
 };
 
-/// Policy that controls how archiver picks delete candidates
-enum class delete_policy_selector {
-    /// Don't keep segments in S3 if they're deleted localy
-    do_not_keep,
-};
+std::ostream& operator<<(std::ostream& s, const upload_candidate& c);
 
-class upload_policy_base {
+class archival_policy {
 public:
-    /// \brief Generate list of segments that should be uploaded to S3
-    ///
-    /// \param remote is a remote manifest (cached or downloade)
-    /// \param lm is log manager
-    virtual std::optional<manifest>
-    generate_upload_set(const manifest& remote, storage::log_manager& lm) = 0;
-};
+    explicit archival_policy(model::ntp ntp);
 
-class delete_policy_base {
-public:
-    /// \brief Generate list of segments that should be remove from S3
+    /// \brief regurn next upload candidate
     ///
-    /// \param remote is a remote manifest (cached or downloade)
-    /// \param lm is log manager
-    virtual std::optional<manifest>
-    generate_delete_set(const manifest& remote, storage::log_manager& lm) = 0;
-};
-
-/// Archiving policy implementation interface
-class archival_policy_base
-  : public upload_policy_base
-  , public delete_policy_base {
-public:
-    archival_policy_base() = default;
-    virtual ~archival_policy_base() = default;
-    archival_policy_base(const archival_policy_base&) = delete;
-    archival_policy_base(archival_policy_base&&) = delete;
-    archival_policy_base& operator=(const archival_policy_base&) = delete;
-    archival_policy_base& operator=(archival_policy_base&&) = delete;
+    /// \param last_offset is a last uploaded offset
+    /// \param lm is a log manager
+    /// \return initializd struct on success, empty struct on failure
+    upload_candidate
+    get_next_candidate(model::offset last_offset, storage::log_manager& lm);
+
+private:
+    model::ntp _ntp;
```

Comment:
can you add a comment here that says you keep only the `key` to the storage `map` so that you don't hold additional memory resources. this is really good but very subtle (vs keeping a log pointer here)